Signal/src/ViewControllers/OWS2FARegistrationViewController.m

```diff
@@ -0,0 +1,291 @@
+    [self.submitButton autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:underscoreView withOffset:kVSpacing];
+    [self.submitButton autoPinWidthToSuperviewWithMargin:self.hMargin];
+
+    [self updateEnabling];
+}
+
+- (void)updateEnabling
+{
+    [self.submitButton setEnabled:self.hasValidPin];
+}
+
+#pragma mark - UITextFieldDelegate
+
+- (BOOL)textField:(UITextField *)textField
+    shouldChangeCharactersInRange:(NSRange)range
+                replacementString:(NSString *)insertionText
+{
+
+    [ViewControllerUtils ows2FAPINTextField:textField
+              shouldChangeCharactersInRange:range
+                          replacementString:insertionText];
+
+    [self updateEnabling];
+
+    return NO;
+}
+
+#pragma mark - Events
+
+- (void)submitButtonWasPressed
+{
+    OWSAssert(self.hasValidPin);
+
+    [self tryToRegister];
+}
+
+- (BOOL)hasValidPin
+{
+    return self.pinTextfield.text.length >= kMin2FAPinLength;
+}
+
+- (void)tryToRegister
+{
+    OWSAssert(self.hasValidPin);
+    OWSAssert(self.verificationCode.length > 0);
+    NSString *pin = self.pinTextfield.text;
+    OWSAssert(pin.length > 0);
+
+    DDLogInfo(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    __weak OWS2FARegistrationViewController *weakSelf = self;
+
+    [ModalActivityIndicatorViewController
+        presentFromViewController:self
+                        canCancel:NO
+                  backgroundBlock:^(ModalActivityIndicatorViewController *modalActivityIndicator) {
+                      OWSProdInfo([OWSAnalyticsEvents registrationRegisteringCode]);
+                      [self.accountManager registerWithVerificationCode:self.verificationCode pin:pin]
+                          .then(^{
+                              OWSAssertIsOnMainThread();
+                              OWSProdInfo([OWSAnalyticsEvents registrationRegisteringSubmittedCode]);
+
+                              DDLogInfo(@"%@ Successfully registered Signal account.", weakSelf.logTag);
+                              dispatch_async(dispatch_get_main_queue(), ^{
+                                  [modalActivityIndicator dismissWithCompletion:^{
+                                      OWSAssertIsOnMainThread();
+
+                                      [weakSelf vericationWasCompleted];
+                                  }];
+                              });
+                          })
+                          .catch(^(NSError *error) {
+                              OWSAssertIsOnMainThread();
+                              OWSProdInfo([OWSAnalyticsEvents registrationRegistrationFailed]);
+                              DDLogError(@"%@ error verifying challenge: %@", weakSelf.logTag, error);
+                              dispatch_async(dispatch_get_main_queue(), ^{
+                                  [modalActivityIndicator dismissWithCompletion:^{
+                                      OWSAssertIsOnMainThread();
+
+                                      [OWSAlerts
```

Comment:
What if the error is due to something besides wrong PIN e.g. rate limit? or network is down? 

Since we're not getting much value out of the generic error title anyway, what about something like:

```
[OWSAlerts showAlertWithTitle:NSLocalizedString(@"REGISTER_2FA_REGISTRATION_FAILED", ...)
      message:error.localizedString];
```

Reply:
Good idea, thanks.